src/compiler/checker.ts

```diff
@@ -4387,48 +4457,50 @@ namespace ts {
                     }
                 }
                 let saveErrorInfo = errorInfo;
-                if (source.flags & TypeFlags.Union || target.flags & TypeFlags.Union) {
-                    if (relation === identityRelation) {
-                        if (source.flags & TypeFlags.Union && target.flags & TypeFlags.Union) {
-                            if (result = unionTypeRelatedToUnionType(<UnionType>source, <UnionType>target)) {
-                                if (result &= unionTypeRelatedToUnionType(<UnionType>target, <UnionType>source)) {
-                                    return result;
-                                }
-                            }
-                        }
-                        else if (source.flags & TypeFlags.Union) {
-                            if (result = unionTypeRelatedToType(<UnionType>source, target, reportErrors)) {
-                                return result;
-                            }
+                if (source.flags & TypeFlags.Reference && target.flags & TypeFlags.Reference && (<TypeReference>source).target === (<TypeReference>target).target) {
+                    // We have type references to same target type, see if relationship holds for all type arguments
+                    if (result = typesRelatedTo((<TypeReference>source).typeArguments, (<TypeReference>target).typeArguments, reportErrors)) {
+                        return result;
+                    }
+                }
+                else if (source.flags & TypeFlags.TypeParameter && target.flags & TypeFlags.TypeParameter) {
+                    if (result = typeParameterRelatedTo(<TypeParameter>source, <TypeParameter>target, reportErrors)) {
+                        return result;
+                    }
+                }
+                else if (relation !== identityRelation) {
+                    if (source.flags & TypeFlags.Union) {
+                        if (result = eachTypeRelatedToType(<UnionType>source, target, reportErrors)) {
+                            return result;
                         }
-                        else {
-                            if (result = unionTypeRelatedToType(<UnionType>target, source, reportErrors)) {
-                                return result;
-                            }
+                    }
+                    else if (target.flags & TypeFlags.Intersection) {
+                        if (result = typeRelatedToEachType(source, <IntersectionType>target, reportErrors)) {
+                            return result;
                         }
                     }
                     else {

```

Comment:
Seems very important that in all cases, the "each" calls come before the "some" calls. I'd add brief comment pointing this out.